### PR TITLE
refactor lazy init to device-agnostic

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* #118091
* #117734
* #117619
* #117611
* #116869
* __->__ #118846

# Motivation
This PR intends to extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for CUDA while maintaining scalability.

# Design
We maintain a flag for each backend to manage the lazy initialization state separately.

# Additional Context
No need more UTs.

### CODE REVIEW

**Comment by guangyey** [comment]
Hi @malfet, I seperate  `device_lazy_init` to this PR. Please help review.

File: torch/csrc/utils/device_lazy_init.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <torch/csrc/Exceptions.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/utils/object_ptr.h>
+#include <iostream>
+namespace torch::utils {
+namespace {
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables,modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
+bool is_initialized[at::COMPILE_TIME_MAX_DEVICE_TYPES] = {false};
```
Comment (malfet) [inline torch/csrc/utils/device_lazy_init.cpp:11]: Please use std::array here, as the construct you wrote here will initialize only first element of the array, see https://godbolt.org
```suggestion
std::array<bool, at::COMPILE_TIME_MAX_DEVICE_TYPES>  is_initialized{};
```

**Review by malfet — APPROVED** [review]
LGTM, but please fix `is_initialized` initialization

File: torch/csrc/utils/device_lazy_init.cpp (hunk 1 above)
Comment (guangyey) [reply torch/csrc/utils/device_lazy_init.cpp:11]: Thanks for your suggestion. Done.

**Comment by guangyey** [comment]
@torchbot merge

**Pull request closed without merging**